### ISSUE TITLE
Provide native support for integration tests setup and teardown

### ISSUE THREAD

**Issue by NikolayMetchev** [gradle/gradle#14111]
It seems to me that Gradle's support for setting up and tearing down of integration testing environments could be improved.

I posted in slack a question regarding integration tests but the solutions provided there were not ideal
https://gradle-community.slack.com/archives/CA7UM03V3/p1597076238216900

Here is a list of features I would like to see that I don't think are possible right now with Gradle.

* Have tasks defined as "setUp" and "tearDown" task for the test task. 
* These tasks should only be run if the test task itself is out of date and not in cache. At the moment finalizedBy runs regardless.
* The tearDown should ideally run even if Gradle gets a Ctrl+C. 

The particular use case I have is using docker-compose to do a setUp and tearDown. Ideally it should be possible to set it up such that those are only run when necessary and they always clean up after themselves. Right now that doesn't seem possible to me.

All this is probably possible if I were to write a custom task myself that does all the logic for me, however I want to be able to re-use tasks defined by other plugins for the setup and teardown steps. In particular I want to re-use the docker-compose plugin tasks.

**Comment by melix** [comment]
I think that's almost exactly the goal of the [build service API](https://docs.gradle.org/current/userguide/build_services.html). In your case it's not _exactly_ what you want because you want to rely on an existing plugin which uses _tasks_ to call docker compose, but your service could probably use the classes from the plugin to do the same. Would you mind taking a look and tell us how this works for you?

**Comment by NikolayMetchev** [comment]
Hello @melix,
Thank you for your suggestion. In the slack channel where I originally raised this somebody did suggest that the docker-compose plugin expose a build service. While I think this has the potential to solve some of the problems I have already outlined it will I think introduce new ones.

I should probably outline what we are doing now:
We have integration tests that require a docker-compose up before the tests run and a docker-compose down when the test finish (regardless if they fail or not).
To achieve this we modified the test task with `dependsOn composeUp` and `finalizedBy composeDown`.
However the story doesn't end there. This setup gives us the following very nice benefits:
A. The docker-compose plugin has a plethora of very nice features that we are taking advantage of and will continue to take advantage of new ones as they are developed. One example is that it probes the docker containers once they are up to wait until they are responsive before actually finishing the docker-compse up task.
B. Our docker-compose.yaml file needs to be actually generated at build time so version numbers can be inject in there amongst other things. This is achieved by a separate Gradle task that we have made a dependency of both the compose up and compose down tasks.
C. The docker images we rely on also get built by Gradle. So we can make those tasks dependencies of the composeUp as well
D. In a local dev scenario we just invoke composeUp on its own and then run individual tests (passing an environment variable that disables the composeUp and composeDown dependancies of the test task)
This actually works very well for us with the only problems being:
1. In CI if the test haven't changed or are cached compose up and compose down still run together with their whole dependency tree
2. Ctrl+C (CI cancel) will leave docker artifacts uncleaned.

If I were to spend time implementing a build service and possibly re-using the docker-compose code I might fix problem 1 but I will probably lose some of the benefits outlined in A-D and I will not fix problem 2.

Overall we are happy with our current situation and are willing to live with the current deficiencies. However if things can be improve for us and everyone in a similar situation to us then great, hence why I put in the feature request in the first place.
It seems to me that this a very common use case in the industry right now and it would be great for Gradle to have a robust solution for integration tests that everyone can just take advantage of.

I haven't looked at the Gradle code at all but an idea I have that might work is if the finalizedBy be enhanced to be able to say "don't run if task is cached". If this feature was added then it could be incorporated to the initalizedBy proposal here: https://github.com/gradle/gradle/issues/11945 and we would be almost there.
